airflow/utils/task_group.py

```diff
@@ -425,6 +426,29 @@ def serialize_for_task_group(self) -> tuple[DagAttributeTypes, Any]:
 
         return DagAttributeTypes.TASK_GROUP, TaskGroupSerialization.serialize_task_group(self)
 
+    def hierarchical_alphabetical_sort(self):
+        """
+        Sorts children in hierarchical alphabetical order:
+        - groups in alphabetical order first
+        - tasks in alphabetical order after them.
+
+        :return: list of tasks in hierarchical alphabetical order
+        """
+
+        def compare_str(a: str, b: str) -> int:
+            return (a > b) - (a < b)
+
+        def compare(a: DAGNode, b: DAGNode) -> int:
+            is_a_group = isinstance(a, TaskGroup)
+            is_b_group = isinstance(b, TaskGroup)
+            if is_a_group == is_b_group:
+                return compare_str(a.node_id, b.node_id)
+            if is_a_group:
+                return -1
+            return 1
+
+        return sorted(self.children.values(), key=cmp_to_key(compare))
```

Comment:
Why not implement a key function? `cmp_to_key` is mainly used for the ease of porting code that was written before key functions were introduced (somewhere like Python 2.5 iirc).

Reply:
The sorting is done by comparing two `DAGNode` objects with each other. This comparison is quite hard to express without hacks (e.g. some negative values for groups) into the single primitive type (`int` or `str`).

If we sort ignore hierarchy (groups first), then simple `key=lambda i: i.node_id` will be better of course.

This part can be re-written without `cmp_to_key` usage but it can be something similar to `cmp_to_key` implementation anyway by a custom object with `>`, `<`, `<=`, `>=` definitions.

Reply:
A key function can return a tuple of primitives as well. I believe returning `(is_group, node_id)` would do the job.